### PR TITLE
Inline structs and add separate instantiate functions

### PR DESCRIPTION
**Pull request by t-sommer**
for the interface types

fixing #667

### CODE REVIEW

File: headers/fmi3FunctionTypes.h [hunk 1]
```diff
@@ -118,41 +84,22 @@ typedef void  (*fmi3CallbackFreeMemory)     (fmi3InstanceEnvironment instanceEnv
...
-                                                      fmi3IntermediateUpdateInfo* intermediateUpdateInfo);
+typedef fmi3Status (*fmi3CallbackIntermediateUpdate) (
+  fmi3InstanceEnvironment instanceEnvironment,
+  fmi3Float64 intermediateUpdateTime,
+  fmi3Boolean eventOccurred,
+  fmi3Boolean clocksTicked,
+  fmi3Boolean intermediateVariableSetAllowed,
+  fmi3Boolean intermediateVariableGetAllowed,
+  fmi3Boolean intermediateStepFinished,
+  fmi3Boolean *canReturnEarly);
```
Comment (IZacharias) [inline headers/fmi3FunctionTypes.h]: canReturnEarly should not be a pointer

File: docs/2_4_common_co-simulation.adoc [hunk 2]
```diff
@@ -245,22 +238,22 @@ _Therefore it is recommended that the FMU provides logging information to the us
...
+If the Co-Simulation master signals the support for intermediate variable access and an FMU has at least one variable with <<intermediateAccess>> = `true`, the FMU can use the callback function <<fmi3CallbackIntermediateUpdate>> to communicate information back to the Co-Simulation master.
 
 [source, C]
 ----
-include::../headers/fmi3FunctionTypes.h[tag=IntermediateUpdateInfo]
+include::../headers/fmi3FunctionTypes.h[tag=CallbackIntermediateUpdate]
 ----
 
-The following flags in the `fmi3IntermediateInfo` struct are used for intermediate variable access:
+The following output arguments in <<fmi3CallbackIntermediateUpdate>> are used for intermediate variable access:
```
Comment (IZacharias) [inline docs/2_4_common_co-simulation.adoc]: The arguments are no output arguments of the function, but input arguments.
The word "output" should be deleted.

File: docs/5_2_hybrid_co-simulation_api.adoc [hunk 3]
```diff
@@ -70,25 +68,25 @@ The Co-Simulation master signals to the FMU that it supports and has recognized
...
-- when `terminateSimulation = true` it means that the model wants to stop integration and the Co-Simulation master should call <<fmi3Terminate>>.
+- When `terminateSimulation = fmi3True`, the model requested to stop integration and the Co-Simulation master should call <<fmi3Terminate>>.
 
-In *Event Mode* it is allowed to `fmi3Get{VariableType}` variable values of the FMU after the call of <<fmi3NewDiscreteStates>> and it is allowed to `fmi3Set{VariableType}` variable values before calling <<fmi3NewDiscreteStates>>.
-The FMU leaves *Event Mode* when the master calls <<fmi3EnterStepMode>> for that FMU.
+In *Event Mode* it is allowed to call `fmi3Get{VariableType}` after <<fmi3NewDiscreteStates>> has been called and it is allowed to call `fmi3Set{VariableType}` before calling <<fmi3NewDiscreteStates>>.
+The FMU leaves *Event Mode* when the master calls <<fmi3EnterStepMode>>.
 
-It is not allowed to call <<fmi3EnterEventMode>> or <<fmi3EnterStepMode>> for an FMU with interface `fmi3CoSimulation` or `fmi3ScheduledCoSimulation`.
+It is not allowed to call <<fmi3EnterEventMode>> or <<fmi3EnterStepMode>> for an FMU with interface type Basic Co-Simulation and Scheduled Co-Simulation.
```
Comment (APillekeit) [inline docs/5_2_hybrid_co-simulation_api.adoc]: what was the reason to switch from or to and?

File: docs/2_4_common_co-simulation.adoc [hunk 4]
```diff
@@ -196,32 +196,25 @@ The Co-Simulation master signals to the FMU which features shall be used for the
...
-|`fmi3CoSimulation`
+|Basic Co-Simulation
 |`canNotUseBasicCoSimulation = false`
 
-|`fmi3HybridCoSimulation`
+|Hybrid Co-Simulation
 |`providesHybridCoSimulation = true`, `canReturnEarlyAfterIntermediateUpdate = true`
 
-|`fmi3ScheduledCoSimulation`
+|fmi3ScheduledCoSimulation
```
Comment (APillekeit) [inline docs/2_4_common_co-simulation.adoc]: fmi3ScheduledCoSimulation --> Scheduled Co-Simulation

File: docs/5_2_hybrid_co-simulation_api.adoc [hunk 5]
```diff
@@ -9,13 +9,11 @@ The <<fmi3CallbackIntermediateUpdate>> callback allows internal events (e.g. ass
 If `fmu3Type` is set to `fmi3HybridCoSimulation`, the <<fmi3CallbackIntermediateUpdate>> callback must be defined in <<fmi3CallbackFunctions>> in <<fmi3Instantiate>>.
 In other cases a NULL pointer can be assigned to the <<fmi3CallbackIntermediateUpdate>> callback, such as if the `fmu3Type` is set to  `fmi3ModelExchange`.
 
-The <<fmi3IntermediateUpdateInfo>>  argument in the <<fmi3CallbackIntermediateUpdate>> function is a structure which is used to signal <<outputClock>> ticks and internal events to the master.
-Whenever the FMU invokes the <<fmi3CallbackIntermediateUpdate>> callback function, it fills the <<fmi3IntermediateUpdateInfo>> structure to pass information concerning the call to the Co-Simulation master.
-The structure is defined as follows:
+The output arguments of <<fmi3CallbackIntermediateUpdate>> are used to signal <<outputClock>> ticks and internal events to the master.
```
Comment (IZacharias) [inline docs/5_2_hybrid_co-simulation_api.adoc]: The word "output" should be deleted.

File: docs/5_2_hybrid_co-simulation_api.adoc [hunk 6]
```diff
@@ -47,7 +45,7 @@ If `canReturnEarly` is `fmi3False` the FMU will not do the early return, regardl
 Note that only the first discontinuity event at a Newtonian time instant shall be signaled that way.
 But in *Event Mode*, there may be an event iteration at a Newtonian time instant and have super dense time instants.
 
-Based on the information delivered in <<fmi3IntermediateUpdateInfo>>, additional information about the discontinuity at that time instant can be obtained from the <<fmi3EventInfo>> structure after calling  <<fmi3NewDiscreteStates>>  and/or <<fmi3GetClock>>.
+Based on the information provided by in <<fmi3CallbackIntermediateUpdate>>, additional information about the discontinuity at that time instant can be obtained by calling <<fmi3NewDiscreteStates>> and <<fmi3GetClock>>.
 
 The FMU must not call the <<fmi3CallbackIntermediateUpdate>> callback function with a `intermediateUpdateInfo.intermediateUpdateTime` value that is smaller than the `intermediateUpdateInfo.intermediateUpdateTime` value given in a previous call of <<fmi3CallbackIntermediateUpdate>> with `intermediateUpdateInfo.intermediateStepFinished` is `fmi3True`.
```
Comment (APillekeit) [inline docs/5_2_hybrid_co-simulation_api.adoc]: Multiple occurences in this and other lines of `intermediateUpdateInfo.intermediateUpdateTime` and `intermediateUpdateInfo.intermediateStepFinished` and other `intermediateUpdateInfo.* need to be changed by removing the intermediateUpdateInfo. part.

File: docs/5_2_hybrid_co-simulation_api.adoc (hunk 3 above)
Comment (t-sommer) [reply docs/5_2_hybrid_co-simulation_api.adoc]: Because it applies to both interface types. Do you want it changed?

**Comment by andreas-junghanns** [comment]
Can I help with this PR? Or do you want to do this all by yourself?

**Comment by t-sommer** [comment]
> Can I help with this PR? Or do you want to do this all by yourself?

Please feel free to make changes (or approve it).

**Comment by t-sommer** [comment]
@andreas-junghanns, I'm updating the Reference FMUs based on the changes in this PR, so we can use them here. Don't waste your time on updating the examples.

**Comment by andreas-junghanns** [comment]
> 
> 
> @andreas-junghanns, I'm updating the Reference FMUs based on the changes in this PR, so we can use them here. Don't waste your time on updating the examples.

Kind of too late: Done here.

File: docs/2_4_common_co-simulation.adoc (hunk 4 above)
Comment (andreas-junghanns) [reply docs/2_4_common_co-simulation.adoc]: This is now fixed with providesScheduledCoSimulation = true

File: docs/5_2_hybrid_co-simulation_api.adoc (hunk 3 above)
Comment (andreas-junghanns) [reply docs/5_2_hybrid_co-simulation_api.adoc]: The second reads better - but it does not link to the section... so far we overdid linking, are we stopping that? I will close this as this is a different issue altogether

File: headers/fmi3FunctionTypes.h [hunk 7]
```diff
@@ -173,14 +120,63 @@ typedef fmi3Status  fmi3SetDebugLoggingTYPE(fmi3Instance instance,
...
+
+typedef fmi3Instance fmi3InstantiateBasicCoSimulationTYPE(
+    fmi3String                     instanceName,
+    fmi3String                     instantiationToken,
+    fmi3String                     resourceLocation,
+    fmi3Boolean                    visible,
+    fmi3Boolean                    loggingOn,
+    fmi3Boolean                    intermediateVariableGetRequired,
+    fmi3Boolean                    intermediateInternalVariableGetRequired,
+    fmi3Boolean                    intermediateVariableSetRequired,
```
Comment (friedrichatgc) [inline headers/fmi3FunctionTypes.h]: Is intermediate access supported by basic co-simulation?

Comment (andreas-junghanns) [reply headers/fmi3FunctionTypes.h]: I would go further: Should Basic Co-Simulation support Intermediate Variable Access?
Or: should IAV be allowed only for Hybrid CS?

Comment (friedrichatgc) [reply headers/fmi3FunctionTypes.h]: See related discussions in issue #820 and #797.

Comment (chrbertsch) [reply headers/fmi3FunctionTypes.h]: > I would go further: Should Basic Co-Simulation support Intermediate Variable Access?
> Or: should IAV be allowed only for Hybrid CS?

I moved this discussion regarding IVA and Early return (from #797, #820, #821) to #836.

**Comment by andreas-junghanns** [comment]
Merging this PR will break the Reference FMUs - Torsten started working on those. I think both merges should happen at the same time.

**Comment by andreas-junghanns** [comment]
> 
> 
> Merging this PR will break the Reference FMUs - Torsten started working on those. I think both merges should happen at the same time.

Just saw: TorstenS already started an a PR is waiting that addresses this inlining of structures:
https://github.com/modelica/Reference-FMUs/pull/9/files

**Comment by andreas-junghanns** [comment]
@chrbertsch: TorstenS is on vacation: should we still merge this PR and https://github.com/modelica/Reference-FMUs/pull/9/files to move forward?

**Comment by chrbertsch** [comment]
> 
> 
> @chrbertsch: TorstenS is on vacation: should we still merge this PR and https://github.com/modelica/Reference-FMUs/pull/9/files to move forward?

I have not had time to look into this. This PR is still marked as a draft PR.
The checks of https://github.com/modelica/Reference-FMUs/pull/9#partial-pull-merging still fail, so I would not yet  merge.
Someone has to fix an review this first.

**Pull request merged**